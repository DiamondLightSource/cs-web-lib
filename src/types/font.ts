import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
