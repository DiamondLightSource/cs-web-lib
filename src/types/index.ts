export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
