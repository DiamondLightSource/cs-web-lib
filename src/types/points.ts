export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
