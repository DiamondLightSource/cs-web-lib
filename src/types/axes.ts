import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
