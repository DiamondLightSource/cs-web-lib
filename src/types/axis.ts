import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
