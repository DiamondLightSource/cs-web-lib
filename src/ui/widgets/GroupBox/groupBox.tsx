import React, { CSSProperties } from "react";

import { Widget } from "../widget";
import { WidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  StringProp,
  ChildrenPropOpt,
  InferWidgetProps,
  ColorPropOpt,
  BoolPropOpt,
  FontPropOpt,
  IntPropOpt
} from "../propTypes";
import { Font } from "../../../types/font";
import { Color } from "../../../types/color";
import Box from "@mui/material/Box";

const INNER_DIV_STYLE: CSSProperties = {
  position: "relative",
  overflow: "visible"
};

const GroupBoxProps = {
  name: StringProp,
  children: ChildrenPropOpt,
  backgroundColor: ColorPropOpt,
  foregroundColor: ColorPropOpt,
  font: FontPropOpt,
  styleOpt: IntPropOpt,
  transparent: BoolPropOpt
};

// Widget that renders a group-box style border showing the name prop.
// This could be replaced if we can implement this as part of the
// border prop.
export const GroupBoxComponent = (
  props: InferWidgetProps<typeof GroupBoxProps>
): JSX.Element => {
  const {
    backgroundColor = Color.fromRgba(240, 240, 240),
    foregroundColor = Color.fromRgba(0, 0, 0),
    font = new Font(16),
    styleOpt = 0,
    transparent = false
  } = props;

  const outerDivStyle: CSSProperties = {
    width: "100%",
    height: "100%",
    position: "absolute",
    padding: "0px",
    boxSizing: "border-box"
  };

  const boxStyle: CSSProperties = {
    width: "100%",
    height: "100%",
    padding: "0px",
    border: "1px solid black",
    whiteSpace: "nowrap",
    overflow: "visible",
    backgroundColor: transparent ? "transparent" : backgroundColor.toString(),
    color: foregroundColor.toString(),
    ...font.css()
  };

  if (styleOpt === 0) {
    // Typical group box with label
    outerDivStyle.padding = "10px";
    outerDivStyle.paddingTop = "0px";
    boxStyle.paddingLeft = "8px";
  } else if (styleOpt === 3) {
    // No groupbox
    boxStyle.border = "none";
  }

  return (
    <div style={outerDivStyle}>
      <Box component="fieldset" sx={boxStyle}>
        {styleOpt === 1 ? (
          <div
            style={{
              height: "20px",
              width: "100%",
              backgroundColor: foregroundColor.toString(),
              ...font.css(),
              color: backgroundColor.toString()
            }}
          >
            {props.name}
          </div>
        ) : (
          <></>
        )}
        {styleOpt === 0 ? <legend>{props.name}</legend> : <></>}
        <div style={INNER_DIV_STYLE}>{props.children}</div>
      </Box>
    </div>
  );
};

const GroupBoxWidgetProps = {
  ...WidgetPropType,
  ...GroupBoxProps,
  name: StringProp,
  children: ChildrenPropOpt
};

export const GroupBox = (
  props: InferWidgetProps<typeof GroupBoxWidgetProps>
): JSX.Element => <Widget baseWidget={GroupBoxComponent} {...props} />;

registerWidget(GroupBox, GroupBoxWidgetProps, "groupbox");
