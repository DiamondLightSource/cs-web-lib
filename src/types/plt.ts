import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
