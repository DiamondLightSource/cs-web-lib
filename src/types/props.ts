import { Color } from "./color";
import { Font } from "./font";
import { MacroMap } from "./macros";
import { WidgetActions } from "../ui/widgets/widgetActions";
import { Border } from "./border";
import { Position } from "./position";
import { PV } from "./pv";
import { Trace } from "./trace";
import { Axes, Axis } from "./axis";
import { Points } from "./points";

export type GenericProp =
  | string
  | string[]
  | boolean
  | number
  | PV
  | Color
  | Font
  | Border
  | Position
  | Rule[]
  | MacroMap
  | WidgetActions
  | OpiFile
  | Trace[]
  | Axes
  | Axis
  | Points;

export interface Expression {
  boolExp: string;
  value: string;
  convertedValue?: GenericProp;
}

interface RulePV {
  pvName: PV;
  trigger: boolean;
}

export interface Rule {
  name: string;
  prop: string;
  outExp: boolean;
  pvs: RulePV[];
  expressions: Expression[];
}

export interface OpiFile {
  path: string;
  macros: MacroMap;
  defaultProtocol: string;
}
