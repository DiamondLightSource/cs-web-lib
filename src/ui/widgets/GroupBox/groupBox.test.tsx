import React from "react";
import { GroupBoxComponent } from "./groupBox";
import { Color } from "../../../types/color";
import { render } from "@testing-library/react";

describe("<GroupBoxComponent /> snapshots", (): void => {
  test("it matches the snapshot for Group Box style", (): void => {
    const { asFragment } = render(
      <GroupBoxComponent
        name={"Test"}
        backgroundColor={Color.WHITE}
        styleOpt={0}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("it matches the snapshot for Title Bar style", (): void => {
    const { asFragment } = render(
      <GroupBoxComponent name={"Title"} styleOpt={1} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("it matches the snapshot for Line style", (): void => {
    const { asFragment } = render(
      <GroupBoxComponent name={"No Title"} styleOpt={2} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("it matches the snapshot for no style", (): void => {
    const { asFragment } = render(
      <GroupBoxComponent name={"None"} styleOpt={3} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("<GroupBoxComponent />", (): void => {
  test("it renders the title", (): void => {
    const grouping = <GroupBoxComponent name={"Test"} styleOpt={1} />;
    const { getByText } = render(grouping);
    expect(getByText("Test")).toBeInTheDocument();
  });

  test("it renders child div with text", (): void => {
    const childText = "Testing Child Component";
    const groupingWithChild = (
      <GroupBoxComponent name={"Test"}>
        <div>{childText}</div>
      </GroupBoxComponent>
    );
    const { getByText } = render(groupingWithChild);
    expect(getByText("Test")).toBeInTheDocument();
  });
});
