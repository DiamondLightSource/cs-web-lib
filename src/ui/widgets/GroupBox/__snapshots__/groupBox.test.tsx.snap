// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<GroupingContainerComponent /> > it matches the snapshot 1`] = `
<div
  style={
    {
      "backgroundColor": "transparent",
      "height": "100%",
      "outline": "1px dotted black",
      "outlineOffset": "-7px",
      "width": "100%",
    }
  }
>
  <div
    style={
      {
        "backgroundColor": "rgba(255,255,255,255)",
        "fontSize": "13px",
        "left": "20px",
        "padding": "0 2px 0 2px",
        "position": "absolute",
        "top": "0",
      }
    }
  >
    Test
  </div>
  <div
    style={
      {
        "padding": "16px",
        "position": "relative",
      }
    }
  />
</div>
`;
